File: test_value_page.mjs
/* 값 검토 페이지의 결정들에 대한 시나리오.
 *
 *     node test_value_page.mjs
 */
import assert from 'node:assert/strict';
import { createRequire } from 'node:module';
const L = createRequire(import.meta.url)('./value_page.js');

let pass = 0; const fails = [];
function test(name, fn) {
  try { fn(); console.log('ok    ' + name); pass++; }
  catch (e) { console.log('FAIL  ' + name + '  <- ' + e.message); fails.push(name); }
}

const NEED = ['Marks_Checked', 'Axis_Labels_Checked', 'Calibration_Checked'];
function state(over) {
  return Object.assign({
    decision: 'APPROVED', who: 'RV_1', subject: 'abc123', panel: 'P_GP001',
    note: '', required: NEED,
    checks: { Marks_Checked: true, Axis_Labels_Checked: true,
              Calibration_Checked: true }}, over || {});
}

test('다 갖춘 승인은 한 줄이 된다', () => {
  const got = L.reviewOf('P_GP001', state());
  assert.equal(got.ready, true);
  assert.equal(got.row.Decision, 'APPROVED');
  assert.equal(got.row.Marks_Checked, 'TRUE');
  assert.equal(got.row.Review_Subject_SHA256, 'abc123');
});

/* REVERT: 확인 칸을 보지 않고 승인으로 친다. `Decision`은 "동의한다"이고 확인
 * 칸은 "무엇을 보았는가"입니다 - 서로 다른 주장이고, 하나로 묶으면 보지 않은
 * 것에 동의한 줄이 생깁니다. `finalize_batch`가 그 줄을 되돌려 보내는데,
 * 그때는 사람이 이미 페이지를 떠난 뒤입니다. */
test('확인이 남아 있으면 승인이 아니다', () => {
  const got = L.reviewOf('P_GP001', state({ checks: { Marks_Checked: true } }));
  assert.equal(got.ready, false);
  assert.match(got.why, /Axis_Labels_Checked/);
});
test('무엇이 남았는지 이름을 댄다', () => {
  assert.match(L.reviewOf('P_GP001', state({ checks: {} })).why, /Marks_Checked/);
});

/* REVERT: 거절과 보류에도 확인 칸을 요구한다. 잘못 앉은 표시를 자세히
 * 확인해 달라고 사람을 붙잡아 두는 일입니다. */
test('거절과 보류는 확인 칸을 묻지 않는다', () => {
  assert.equal(L.reviewOf('P_GP001',
    state({ decision: 'REJECTED', checks: {} })).ready, true);
  assert.equal(L.reviewOf('P_GP001',
    state({ decision: 'HOLD', checks: {} })).ready, true);
});
test('거절이 확인하지 않은 것을 확인했다고 적지 않는다', () => {
  const got = L.reviewOf('P_GP001', state({ decision: 'REJECTED', checks: {} }));
  assert.equal(got.row.Marks_Checked, 'FALSE');
});

/* REVERT: 이름 없이도 판정으로 친다. `finalize_batch`는 등록된 사람만 승인할
 * 수 있다고 하고, 이름 없는 줄은 그 확인을 받을 수조차 없습니다. */
test('누가 보았는지 없으면 판정이 아니다', () => {
  const got = L.reviewOf('P_GP001', state({ who: '  ' }));
  assert.equal(got.ready, false);
  assert.match(got.why, /누가 보았는지/);
});

/* REVERT: 실행 지문 없이도 답으로 친다. 그러면 이 답이 어느 추출에 대한
 * 것인지 알 수 없고, 다른 실행의 승인이 이 실행에 상속됩니다. */
test('실행 지문이 없으면 답이 아니다', () => {
  const got = L.reviewOf('P_GP001', state({ subject: '' }));
  assert.equal(got.ready, false);
  assert.match(got.why, /지문/);
});

test('아직 고르지 않은 줄은 답이 아니다', () => {
  assert.match(L.reviewOf('P_GP001', state({ decision: '' })).why, /아직 고르지/);
});
test('받을 수 없는 답은 이름을 대며 거절한다', () => {
  assert.match(L.reviewOf('P_GP001', state({ decision: 'MAYBE' })).why, /MAYBE/);
});
test('묻지 않은 확인은 빈 칸으로 나간다', () => {
  assert.equal(L.reviewOf('P_GP001', state()).row.Identity_Checked, '');
});

/* REVERT: 본 날짜를 브라우저 시계에서 가져온다. 이 프로그램이 아는 것이
 * 아니고, 아는 척하면 나중에 그 날짜가 증거가 됩니다. */
test('본 날짜는 부른 쪽이 준다', () => {
  const csv = L.buildCsv(['P_GP001'], { P_GP001: state() }, '2026-09-10');
  assert.match(csv, /2026-09-10/);
});
/* REVERT: 오늘 날짜를 브라우저에서 가져온다. 이 프로그램이 아는 것이 아니고,
 * 아는 척하면 나중에 그 날짜가 "언제 보았는가"의 증거가 됩니다. */
test('부르는 쪽이 날짜를 주지 않으면 날짜도 없다', () => {
  const csv = L.buildCsv(['P_A'], { P_A: state({ panel: 'P_A' }) }, '');
  assert.equal(csv.split('\n')[1].split(',')[10], '""');
});
/* REVERT: 화면의 열쇠를 그대로 Panel_ID로 내보낸다. 화면이 패널 하나를 두 번
 * 싣거나 이름에 무언가를 붙여 두면, 나가는 이름이 실행의 패널 이름이 아닙니다. */
test('패널 이름은 상태가 들고 있는 것을 쓴다', () => {
  assert.equal(L.reviewOf('key', state({ panel: 'P_GP007' })).row.Panel_ID,
               'P_GP007');
});
/* REVERT: 목록이 아니라 상태에 있는 이름을 전부 내보낸다. 브라우저에 남아
 * 있던 지난 실행의 답이 이 실행의 답안지에 섞여 나갑니다. */
test('csv는 목록에 있는 패널만 싣는다', () => {
  assert.equal(L.buildCsv(['P_A'],
    { P_A: state({ panel: 'P_A' }), P_OLD: state({ panel: 'P_OLD' }) }, '')
    .split('\n').length, 2);
});
test('csv는 답이 된 줄만 싣고 번호를 다시 매긴다', () => {
  const csv = L.buildCsv(['P_A', 'P_B'],
    { P_A: state({ panel: 'P_A' }), P_B: state({ decision: '' }) }, '2026-09-10');
  const lines = csv.split('\n');
  assert.equal(lines.length, 2);
  assert.match(lines[1], /"R001"/);
});
test('따옴표가 든 메모는 csv를 깨뜨리지 않는다', () => {
  assert.match(L.buildCsv(['P_A'],
    { P_A: state({ panel: 'P_A', note: '표시가 "옆"에 앉음' }) }, ''),
    /"표시가 ""옆""에 앉음"/);
});
test('남은 일과 보류를 따로 센다', () => {
  const st = { a: state({ decision: 'HOLD', checks: {} }), b: state({ decision: '' }) };
  assert.equal(L.remaining(['a', 'b'], st), 1);
  assert.equal(L.held(['a', 'b'], st), 1);
});

console.log('');
console.log(pass + '/' + (pass + fails.length) + ' passed');
if (fails.length) { console.log('FAILED: ' + fails.join(', ')); process.exit(1); }

File: test_value_page.py
# -*- coding: utf-8 -*-
"""값 검토 페이지가 사람에게 무엇을 내미는가.

    python3 test_value_page.py     # exit 0 = all scenarios pass

무엇이 답이 되는지는 `value_page.js`가 정하고 `test_value_page.mjs`가 봅니다.
여기서 붙잡는 것은 화면의 성질입니다: 겹쳐 그린 그림이 실려 나가는가, 실행
지문이 화면의 칸이 아니라 논리로 건너가는가, 확인 칸이 이름이 아니라 물음으로
적히는가, 그리고 내려받는 이름이 `finalize_batch`가 읽는 파일과 다른가.
"""
import csv
import io
import os
import re
import shutil
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)

import value_page as V                                           # noqa: E402

N = [0]
FAIL = []


def check(name, ok, detail=""):
    N[0] += 1
    print("  %s %s%s" % ("ok  " if ok else "FAIL", name,
                         "" if ok else "  <- %s" % (detail,)))
    if not ok:
        FAIL.append(name)


TMP = tempfile.mkdtemp(prefix="fdt-value-page-")
VALS = os.path.join(TMP, "values")
os.makedirs(VALS)

COLUMNS = ("Panel", "Outcome", "Group", "Median", "Q1", "Q3",
           "Whisker_Lower", "Whisker_Upper")
ROWS = [
    dict(Panel="P_A", Outcome="Plasma total calcium (mmol/L)", Group="BDC",
         Median="2.354", Q1="2.328", Q3="2.412",
         Whisker_Lower="2.248", Whisker_Upper="2.461"),
    dict(Panel="P_A", Outcome="Plasma total calcium (mmol/L)", Group="D2",
         Median="2.429", Q1="2.373", Q3="2.468",
         Whisker_Lower="2.298", Whisker_Upper="2.511"),
    dict(Panel="P_B", Outcome="Serum CTx (pmol/L)", Group="BDC",
         Median="4389", Q1="3387", Q3="6465",
         Whisker_Lower="1237", Whisker_Upper="12110"),
]
with io.open(os.path.join(VALS, V.VALUES), "w", encoding="utf-8", newline="") as fh:
    w = csv.DictWriter(fh, fieldnames=list(COLUMNS))
    w.writeheader()
    w.writerows(ROWS)
#: 겹쳐 그린 그림은 한 패널에만 둡니다. 없는 그림이 어떻게 나가는지도 이
#: 페이지의 성질입니다 - 볼 것이 없는 카드는 확인을 청하는 카드가 아닙니다.
with io.open(os.path.join(VALS, "P_A.png"), "wb") as fh:
    fh.write(b"\x89PNG\r\n\x1a\n" + b"\x00" * 8)

SUBJECTS = {"P_A": "abc123"}
HTML, COUNT = V.build(VALS, when="2026-09-10", subjects=SUBJECTS,
                      log=lambda *a: None)

print("패널마다 한 장, 값은 표로")
check("패널마다 카드가 하나", COUNT == 2 and HTML.count("class='doc'") == 2,
      "%s / %s" % (COUNT, HTML.count("class='doc'")))
check("그림이 있으면 실린다", "data:image/png;base64," in HTML)
# REVERT: 겹쳐 그린 그림이 없어도 조용히 넘어간다. 볼 것이 없는 카드는 확인을
# 청하는 카드가 아니고, 사람은 무엇을 보고 답해야 할지 모릅니다.
check("그림이 없으면 없다고 말한다", "겹쳐 그린 그림 없음" in HTML)
check("읽힌 값이 표로 함께 보인다",
      "2.354" in HTML and "4389" in HTML and "수염 아래" in HTML)

print()
print("실행 지문은 화면의 칸이 아니라 논리로 건너간다")
# REVERT: 실행 지문을 입력 칸으로 내민다. 사람이 고칠 수 있어서는 안 되는
# 값이고, 고쳐지면 이 승인이 다른 실행의 값에 붙습니다.
check("지문은 META로만 간다",
      ('"subject": "abc123"' in HTML or '"subject":"abc123"' in HTML)
      and "data-subject" not in HTML)
# REVERT: 지문이 없는 패널도 있는 것처럼 내민다. 그 카드의 답은
# `finalize_batch`가 어느 추출에 대한 것인지 알 수 없습니다.
check("지문이 없는 패널은 없다고 적힌다", "실행 지문 없음" in HTML)

print()
print("확인 칸은 이름이 아니라 물음으로 적힌다")
# REVERT: 확인 칸에 열 이름만 적는다. `Axis_Labels_Checked`라고만 쓰여 있으면
# 무엇을 확인하라는 것인지 알 수 없고, 모르면 누르는 것이 확인이 아닙니다.
for _name, _ask in V.ASKS:
    check("  %s → %s" % (_name, _ask[:22]), _ask in HTML, _name)
# REVERT: 무엇을 확인해야 하는지 논리에 넘겨주지 않는다. 그러면 확인 칸이
# 화면에는 있는데 아무것도 요구하지 않고, 하나도 누르지 않은 승인이 그대로
# 답이 됩니다 - 이 페이지가 있는 까닭이 사라집니다.
_need = re.search(r'"required":\s*\[(.*?)\]', HTML, re.S)
check("논리가 무엇을 요구하는지 페이지가 심어 준다",
      _need is not None
      and set(re.findall(r'"([A-Za-z_]+)"', _need.group(1)))
      == set(name for name, _ask in V.ASKS),
      _need.group(1) if _need else "없음")

print()
print("화면과 논리가 같은 어휘를 쓴다")
with io.open(os.path.join(HERE, "value_page.js"), encoding="utf-8") as fh:
    LOGIC = fh.read()
_js = set(re.findall(r"'([A-Z_]+)'",
                     re.search(r"var DECISIONS = \[(.*?)\]", LOGIC, re.S).group(1)))
check("고를 수 있는 답이 곧 답이 되는 답이다",
      set(v for v, _l in V.LABELS) == _js,
      "화면 %s / 논리 %s" % (sorted(v for v, _l in V.LABELS), sorted(_js)))
check("판정 논리가 페이지 안에 들어 있다", "function reviewOf(" in HTML)

print()
print("내려받는 이름은 finalize_batch가 읽는 파일과 다르다")
# REVERT: `value_review.csv`로 내려받는다. 사람이 그 자리에 두게 되고, 실행이
# 만들어 둔 빈 대기열을 덮어씁니다.
check("답안지는 value_review_answers.csv", "'value_review_answers.csv'" in HTML)
check("본 날짜는 부른 쪽이 준 것이 실린다", '"2026-09-10"' in HTML)

print()
print("읽힌 값이 없는 폴더는 페이지가 아니다")
_empty = os.path.join(TMP, "none")
os.makedirs(_empty)
try:
    V.build(_empty, log=lambda *a: None)
    check("값이 없으면 멈추고 말한다", False, "멈추지 않았다")
except SystemExit as exc:
    check("값이 없으면 멈추고 말한다", "읽힌 값이 없습니다" in str(exc), str(exc))

shutil.rmtree(TMP, ignore_errors=True)
print()
print("FDT_SCENARIOS_RUN=%d" % N[0])
print("%d scenarios run" % N[0])
if FAIL:
    print("%d FAILED: %s" % (len(FAIL), FAIL))
    raise SystemExit(1)
print("all scenarios passed")
